src/mono/mono/mini/mini-runtime.c

```diff
@@ -5395,3 +5401,33 @@ mini_alloc_jinfo (MonoJitMemoryManager *jit_mm, int size)
 		return (MonoJitInfo*)mono_mem_manager_alloc0 (jit_mm->mem_manager, size);
 	}
 }
+
+void
+mono_set_runtime_init_callback (MonoRuntimeInitCallback callback)
+{
+	runtime_init_callback = callback;
+}
+
+/*
+* Default implementation invoking runtime init callback in native-to-managed wrapper.
+*/
+void
+mono_invoke_runtime_init_callback (void)
+{
+	if (G_UNLIKELY (runtime_init_callback)) {
+		guint64 thread_id = mono_native_thread_os_id_get ();
+		if (runtime_init_callback && (runtime_init_thread_id == thread_id))
+			return;
+
+		while (mono_atomic_cas_i64 ((volatile gint64 *)&runtime_init_thread_id, (gint64)thread_id, (gint64)G_MAXUINT64) != (gint64)G_MAXUINT64)
+			g_usleep (1000);
```

Comment:
Why do we early return if `mono_native_thread_os_id_get` is the max value? What is `syscall(SYS_gettid)`
?

I'm not sure I'm understanding the function properly, but doesnt the function return the value at `&runtime_init_thread_id` before any operations, and only changes the value at `&runtime_init_thread_id` to `false` (`(gint64)G_MAXUINT64) != (gint64)G_MAXUINT64`) if `thread_id` is the same as the value at `&runtime_init_thread_id`?
But because of the early return, thread_id will not be the `G_MAXUINT64` value, right? Or does `thread_id` somehow get updated after `g_usleep (1000)`?

Reply:
This is a compare and swap loop, it makes sure only one thread can move the value from G_MAXUINT64 to its own thread id, meaning that only one thread gets passed mono_atomic_cas_i64 one at a time, the rest will be blocked in the loop until they managed to exchange runtime_init_thread_id from G_MAXUINT64 to their own thread id, as long as the value is different than G_MAXUINT64 then they will wait for their turn to win the "race".

The early return is done to handle recursive calls into this function by the same thread. When the thread does the callback it will run tons of code, that could in theory trigger a call to a native-to-managed wrapper that in turn would end up here, but in that case check against already set thread id and will see that its a recursive call (runtime_init_thread_id holds its own thread id) and make early exit instead of deadlock on the CAS loop.

Reply:
Ah I missed a parenthesis when first seeing this, I also missed that `mono_atomic_cas_i64` swapped the `comp` and `exch` when passing it to `gcc_sync_val_compare_and_swap` so I had a misunderstanding how the comparison worked. I see how it flows through now, and without the lock, would there ever be the case that two threads execute the initial `mono_atomic_cas_i64 ((volatile gint64 *)&runtime_init_thread_id, (gint64)thread_id, (gint64)G_MAXUINT64)` at the same time or is that not possible?

Reply:
all threads can call mono_atomic_cas_i64 at the same time, but only one will succeed setting its thread value, the others will spin in the loop until the thread value is reset to G_MAXUINT64 and then another thread will succeed its mono_atomic_cas_i64 call and so on until all threads have get passed the CAS loop but since the first thread has reset the callback all other threads will see NULL and return out of the function.